client/src/JobSelector.jsx

```diff
@@ -1,56 +1,57 @@
-import React from "react";
+import React, { useState } from "react";
 import PropTypes from "prop-types";
 
 import Category from "./Category";
 import AjaxHelper from "./AjaxHelper";
 
-class JobSelector extends React.Component {
-    constructor(props) {
-        super(props);
-        this.state = { active: true, data: [] };
-        this.ajax = AjaxHelper.getInstance();
-    }
-
-    hide() {
-        var newState = this.state;
-        newState.active = false;
-        this.setState(newState);
-    }
-
-    show() {
-        this.setState({ active: true, data: [] });
-        this.componentDidMount();
-    }
-
-    componentDidMount() {
-        this.ajax.getJSON(
-            this.ajax.ingestApiUrl,
+const JobSelector = () => {
+    const [active, setActive] = useState(true);
+    const [data, setData] = useState([]);
+    const ajax = AjaxHelper.getInstance();
+
+    const hide = () => {
+        setActive(false);
+    };
+    const show = () => {
+        setActive(true);
+        ajax.getJSON(
+            ajax.ingestApiUrl,
             null,
-            function (data) {
-                this.setState({ active: true, data: data });
-            }.bind(this)
-        );
-    }
-
-    render() {
-        var categories = [];
-        var empty_categories = [];
-        for (var i in this.state.data) {
-            var category = this.state.data[i];
-            var element = <Category key={category.category} data={category} />;
-            if (category.jobs.length > 0) {
-                categories[categories.length] = element;
-            } else {
-                empty_categories[empty_categories.length] = element;
+            (data) => {
+                setActive(true);
+                setData(data);
             }
-        }
-        return (
-            <div className={this.state.active ? "" : "hidden"} id="jobSelector">
-                {categories}
-                {empty_categories}
-            </div>
         );
-    }
+    };
+
+    var categories = [];
+    var empty_categories = [];
+    const categoryComponents = Object.keys(data).reduce((categoryComponents, key) => {
+        const category = data[key];
+        const categoryComponent = <Category key={category.category} data={category} />;
+        if(category.jobs.length > 0) {
+            categoryComponents[0] = [...categoryComponents[0], categoryComponent];
```

Comment:
I wonder if it would be a little less confusing if you used a different name for the `categoryComponents` callback function parameter, and the `categoryComponents` constant that the result of the `reduce` operation is assigned to. The reused name caused me some momentary confusion when reading this code.

Reply:
Also, would `categoryComponents[0].push(categoryComponent)` be functionally equivalent yet more concise, or am a missing a subtle reason why this approach is preferable?

Reply:
sorry. I'm just used to doing this. There are cases where this would be better.